subprojects/core/src/integTest/groovy/org/gradle/api/tasks/BuildCacheSymlinksIntegrationSpec.groovy

```diff
@@ -0,0 +1,243 @@
+        }
+
+        where:
+        linksStrategy              | upToDate
+        LinksStrategy.PRESERVE_ALL | false
+        LinksStrategy.FOLLOW       | false // should be true
+    }
+
+    def "when dir link is replaced by its contents, the task should have up-to-date=#upToDate if linksStrategy=#linksStrategy"() {
+        checkUpToDate(linksStrategy, upToDate) {
+            dirLink.delete()
+            FileUtils.copyDirectory(originalDir, dirLink)
+        }
+
+        where:
+        linksStrategy              | upToDate
+        LinksStrategy.PRESERVE_ALL | false
+        LinksStrategy.FOLLOW       | false // should be true
+    }
+
+    def "when file link is replaced by its contents, the task should have up-to-date=#upToDate if linksStrategy=#linksStrategy"() {
+        checkUpToDate(linksStrategy, upToDate) {
+            fileLink.delete()
+            FileUtils.copyFile(originalFile, fileLink)
+        }
+
+        where:
+        linksStrategy              | upToDate
+        LinksStrategy.PRESERVE_ALL | false
+        LinksStrategy.FOLLOW       | false // should be true
+    }
+
+    def "when file is replaced by a link pointing to the original, the task should have up-to-date=#upToDate if linksStrategy=#linksStrategy"() {
+        fileLink.delete()
+        dirLink.delete()
+        originalDir.createFile(fileLink.name) << originalFile.text
+
+        checkUpToDate(linksStrategy, upToDate) {
+            fileLink.delete()
+            originalDir.file(fileLink.name).createLink(originalFile.getRelativePathFromBase())
+        }
+
+        where:
+        linksStrategy              | upToDate
+        LinksStrategy.PRESERVE_ALL | false
+        LinksStrategy.FOLLOW       | false // should be true
+    }
+
+    def "when configuration is changed, the task should have up-to-date=#upToDate if linksStrategy was #linksStrategy and became #linksStrategyAfter"() {
+        checkUpToDate(linksStrategy, upToDate) {
+            buildKotlinFile.text = """
+            tasks.register<Copy>("cp") {
+                linksStrategy = LinksStrategy.$linksStrategyAfter
+                from("${inputDirectory.name}")
+
+                into("${outputDirectory.name}")
+            }
+            """
+        }
+
+        where:
+        linksStrategy              | linksStrategyAfter         | upToDate
+        LinksStrategy.PRESERVE_ALL | LinksStrategy.FOLLOW       | false
+        LinksStrategy.FOLLOW       | LinksStrategy.PRESERVE_ALL | false
+    }
+
+    def "when a relative link is replaced to absolute, the task should have up-to-date=#upToDate if linksStrategy=#linksStrategy"() {
+        checkUpToDate(linksStrategy, upToDate) {
+            fileLink.delete()
+            Files.createSymbolicLink(fileLink.toPath(), Paths.get(originalFile.getCanonicalPath()))
+        }
+
+        where:
+        linksStrategy              | upToDate
+        LinksStrategy.PRESERVE_ALL | false
```

Comment:
🤔 In this test and the previous ones, we never check the `PRESERVE_RELATIVE` behavior. While the result is the same, it might be a good idea to protect against any unintended change for that specific strategy.